### PR TITLE
feat(platform): add ability to override platform detection methods

### LINKED ISSUE #19737: feat: Add ability to override platform detection methods
# Feature Request

**Ionic version:**

[x] **4.x**

**Describe the Feature Request**

The platform detection methods Ionic provides cover most use cases; however, it is not going to be possible for us to reasonably detect 100% of all devices correctly. This is particularly true for off brand Android devices that do not conform to a standard when it comes to identifying themselves on the web.

It would be good for us to allow users to override our own methods with their custom methods in case they need to handle those specific edge cases in their apps.

**Describe Preferred Solution**

The easiest way to do this is to let developers customize the `PLATFORMS_MAP` object: https://github.com/ionic-team/ionic/blob/master/core/src/utils/platform.ts#L118

I see two options for doing this:

1. Expose a function (tentatively) called `setPlatformDetection(platform: Platform, callback: any)`. This function will allow users to provide a platform and a detection function. Usage is as follows:

```javascript
import { setPlatformDetection } from '@ionic/core';

const customAndroidFunction = (win: Window) => {
  return window.someExample === true;
}

setPlatformDetection('android', customAndroidFunction);
```
(Note: The `any` type will be replaced with the appropriate function type once that is determined)

2. Opt for a config approach where developers specify this on app start. Ionic Angular usage is as follows:

```javascript
...

IonicModule.forRoot({
  'platformDetection': {
    'android': customAndroidFunction
  }
});

...
```
After discussing with the team, I am leaning more towards option 2. I can see things getting a bit messy if users override the platform detection using option 1 _after_ the app has already been initialized. 

**Additional Context**

We might also want to consider exposing the `testUserAgent` and `matchMedia` functions that Ionic uses internally. I could see those being helpful when customizing the platform detection functions.

### PR DESCRIPTION
**Pull request by willmartindev**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Issue Number: resolves #19737 

## What is the new behavior?

- Default platform detection methods can now be overridden by a custom implementation. 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/utils/config.ts [hunk 1]
```diff
@@ -175,6 +177,11 @@ export interface IonicConfig {
    */
   sanitizerEnabled?: boolean;
 
+  /**
+   * Overrides the default platform detection methods.
+   */
+  platformDetection?: PlatformDetectionConfig;
```
Comment (liamdebeasi) [inline core/src/utils/config.ts:183]: Can we call this "platform"? (And also `PlatformConfig`)

**Review by liamdebeasi — CHANGES_REQUESTED** [review]
This is looking great, but we do need to add tests to verify this works. Additionally, we need to update the Config docs in the `ionic-docs` repo to show usage, the `PlatformDetectionConfig` interface, etc.

File: packages/react-router/test-app/src/index.tsx [hunk 2]
```diff
@@ -1,8 +1,19 @@
...
+setupConfig({
+  platform: {
+    'cordova': (win: any) => {
+      const isCordova = !!(win['cordova'] || win['phonegap'] || win['PhoneGap']);
+      console.log('Checking if platform is "cordova" with a custom detection method: ', isCordova);
+      return isCordova;
+    },
+  }
+});
+
```
[the remark this answers is not in the document]
Comment (willmartindev) [reply packages/react-router/test-app/src/index.tsx:16]: No, I can remove it since I added the spec test in core. 👍

**Review by liamdebeasi — APPROVED** [review]
Great job on this!

**Pull request merged**